nameko_stripe/handlers/event.py

```diff
@@ -0,0 +1,131 @@
+import json
+import logging
+from collections import defaultdict
+from fnmatch import fnmatch
+from functools import partial
+
+from eventlet.event import Event
+from nameko.exceptions import serialize
+from nameko.web.server import WebServer
+from werkzeug.exceptions import BadRequest, HTTPException
+from werkzeug.routing import Map, Rule
+from werkzeug.wrappers import Request, Response
+
+
+from nameko_stripe.handlers.webhook import WebhookHandler
+
+
+logger = logging.getLogger(__name__)
+
+
+class Router:
+    def __init__(self, container, handlers=None):
+        self.container = container
+        self.handlers = handlers or set()
+
+    def __call__(self, request):
+        return self.handle(request)
+
+    def handle(self, request):
+        events = set()
+        event_name = self.parse_event_name(request)
+        for handler in self.handlers:
+            if fnmatch(event_name, handler.event_name):
+                event = Event()
+                events.add(event)
+                handler.handle_request(request, event)
+        [event.wait() for event in events]
+
+    def parse_event_name(self, request):
+        data = request.data.decode("utf8")
+        try:
+            return json.loads(data)["type"]
+        except Exception:
+            logger.warning("Cannot parse Stripe webhook", exc_info=True)
+            raise BadRequest("Cannot parse Stripe webhook")
+
+
+class EventServer(WebServer):
+    def make_url_map(self):
+
+        routes = defaultdict(list)
+
+        for provider in self._providers:
+            routes[(provider.url, provider.method)].append(provider)
+
+        url_map = Map()
+
+        for route, providers in routes.items():
+            url, method = route
+            rule = Rule(url, methods=method.split(","))
+            rule.endpoint = Router(self.container, providers)
```

Comment:
Let's try not to couple those classes too much using composition this way.

What about we do the following?

```python
class EventServer(WebServer):

    router_cls = Router
    wsgi_app_cls = WsgiApp

    # ...
```


This will make it easier to extend either `Router` or `WsgiApp` outside this library, if we needed to, and we would just have to do:

```python
class EventServer(EventServerBase):

    router_cls = NewRouter
    wsgi_app_cls = NewWsgiApp


class EventHandler(WebhookHandler):

     server = EventServer()


event_handler = EventHandler.decorator
````